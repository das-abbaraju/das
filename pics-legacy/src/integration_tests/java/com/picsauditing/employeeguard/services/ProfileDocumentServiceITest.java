package com.picsauditing.employeeguard.services;

import com.picsauditing.PICS.I18nCache;
import com.picsauditing.access.Permissions;
import com.picsauditing.jpa.entities.Account;
import com.picsauditing.web.SessionInfoProvider;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import static org.mockito.Mockito.when;

@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration( {"classpath:EGITest-localhost-context.xml"} )
public class ProfileDocumentServiceITest {

	@Mock
	private Permissions permissions;

	@Mock
	private SessionInfoProvider sessionInfoProvider;

	@Autowired
	private ProfileDocumentService profileDocumentService;

	private static final int SKILL_ID = 11;
	private static final int DOCUMENT_ID = 9;
	private static final int APP_USER_ID = 131073;

	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);
	}

	@Test
	public void testUpdate_getDocumentThumbnail_CheckEmployeePermissions() throws Exception {
		when(sessionInfoProvider.getAppUserId()).thenReturn(APP_USER_ID);
		profileDocumentService.getAccountSkillProfileForEmployeeAndSkill(SKILL_ID, DOCUMENT_ID);
	}

}
