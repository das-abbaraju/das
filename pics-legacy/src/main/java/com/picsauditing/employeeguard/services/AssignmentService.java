package com.picsauditing.employeeguard.services;

import com.picsauditing.employeeguard.entities.*;
import com.picsauditing.employeeguard.models.EntityAuditInfo;
import com.picsauditing.employeeguard.services.entity.EmployeeEntityService;
import com.picsauditing.employeeguard.services.entity.ProjectEntityService;
import com.picsauditing.employeeguard.services.entity.RoleEntityService;
import com.picsauditing.employeeguard.services.entity.SkillEntityService;
import com.picsauditing.employeeguard.services.external.AccountService;
import com.picsauditing.employeeguard.services.processor.ProjectAssignmentDataSet;
import com.picsauditing.employeeguard.services.processor.ProjectAssignmentProcess;
import com.picsauditing.employeeguard.services.processor.RoleAssignmentProcess;
import com.picsauditing.employeeguard.util.PicsCollectionUtil;
import org.apache.commons.collections.CollectionUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

public class AssignmentService {

	@Autowired
	private AccountService accountService;
	@Autowired
	private EmployeeEntityService employeeEntityService;
	@Autowired
	private ProjectEntityService projectEntityService;
	@Autowired
	private RoleEntityService roleEntityService;
	@Autowired
	private SkillEntityService skillEntityService;

	public Map<Project, Map<Employee, Set<AccountSkill>>> getEmployeeSkillsForProjectsUnderSite(final int siteId) {

		List<Project> siteProjects = projectEntityService.getAllProjectsForSite(siteId);
		Map<Project, Set<Employee>> projectEmployees = employeeEntityService.getEmployeesByProjects(siteProjects);
		Map<Role, Set<Employee>> roleEmployees = getEmployeesByRole(siteId, siteProjects);
		Map<Project, Set<AccountSkill>> projectRequiredSkill = skillEntityService
				.getRequiredSkillsForProjects(siteProjects);
		Map<Project, Set<Role>> projectRoles = roleEntityService.getRolesForProjects(siteProjects);
		Map<Role, Set<AccountSkill>> projectRoleSkills = skillEntityService
				.getSkillsForRoles(PicsCollectionUtil.mergeCollectionOfCollections(projectRoles.values()));
		Set<AccountSkill> siteAndCorporateRequiredSkills = skillEntityService
				.getSiteRequiredSkills(siteId, accountService.getTopmostCorporateAccountIds(siteId));

		ProjectAssignmentDataSet dataSet = new ProjectAssignmentDataSet.Builder()
				.projectEmployees(projectEmployees)
				.projectRequiredSkills(projectRequiredSkill)
				.projectRoles(projectRoles)
				.projectRoleSkills(projectRoleSkills)
				.projects(siteProjects)
				.roleEmployees(roleEmployees)
				.siteAndCorporateRequiredSkills(siteAndCorporateRequiredSkills)
				.build();

		return new ProjectAssignmentProcess().getProjectSkillsForEmployees(dataSet);
	}

	/**
	 * Retrieves a map of employees to both site and project roles for the specific site
	 *
	 * @param siteId
	 * @param projects
	 * @return
	 */
	public Map<Role, Set<Employee>> getEmployeesByRole(final int siteId, final Collection<Project> projects) {
		return new RoleAssignmentProcess().getCorporateRoleEmployees(
				employeeEntityService.getEmployeesByProjectRoles(projects),
				employeeEntityService.getEmployeesBySiteRoles(Arrays.asList(siteId)));
	}

	public Map<Employee, Set<AccountSkill>> getEmployeeSkillsForSite(final int siteId) {
		Set<Employee> employeesAssignedToSite = getEmployeesAssignedToSite(siteId);

		Map<Employee, Set<AccountSkill>> employeeSkillsForProjects = getEmployeeSkillsForProjects(siteId, employeesAssignedToSite);
		Map<Employee, Set<AccountSkill>> employeeSkillsForRole = getEmployeeSkillsForRoles(siteId);

		Map<Employee, Set<AccountSkill>> allEmployeeRequiredSkills = PicsCollectionUtil.mergeMapOfSets(employeeSkillsForProjects, employeeSkillsForRole);
		List<Integer> accountIdsInHierarchy = accountService.getTopmostCorporateAccountIds(siteId);

		return appendSiteAndCorporateSkills(allEmployeeRequiredSkills,
				skillEntityService.getSiteRequiredSkills(siteId, accountIdsInHierarchy));
	}

	private Map<Employee, Set<AccountSkill>> getEmployeeSkillsForProjects(int siteId, Set<Employee> employeesAssignedToSite) {
		Map<Employee, Set<Project>> employeeProjects = projectEntityService.getProjectsForEmployeesBySiteId(employeesAssignedToSite, siteId);
		Set<Project> allProjects = PicsCollectionUtil.extractAndFlattenValuesFromMap(employeeProjects);
		Map<Project, Set<AccountSkill>> projectSkills = skillEntityService.getRequiredSkillsForProjects(allProjects);
		return getKeyToSetOfValues(employeeProjects, projectSkills);
	}

	private Map<Employee, Set<AccountSkill>> getEmployeeSkillsForRoles(int siteId) {
		Map<Employee, Set<Role>> employeeRoles = getAllEmployeeRolesForSite(siteId);
		Set<Role> allRoles = PicsCollectionUtil.extractAndFlattenValuesFromMap(employeeRoles);
		Map<Role, Set<AccountSkill>> roleSkills = skillEntityService.getSkillsForRoles(allRoles);
		return getKeyToSetOfValues(employeeRoles, roleSkills);
	}

	private <KEY, ENTITY, VALUE> Map<KEY, Set<VALUE>> getKeyToSetOfValues(Map<KEY, Set<ENTITY>> keyToEntities, Map<ENTITY, Set<VALUE>> entityToValues) {
		Map<KEY, Set<VALUE>> keyValues = new HashMap<>();
		for (Map.Entry<KEY, Set<ENTITY>> keyEntityEntry : keyToEntities.entrySet()) {
			KEY key = keyEntityEntry.getKey();

			keyValues.put(key, new HashSet<VALUE>());

			for (ENTITY entity : keyEntityEntry.getValue()) {
				if (entityToValues.containsKey(entity)) {
					keyValues.get(key).addAll(entityToValues.get(entity));
				}
			}
		}

		return keyValues;
	}

	private <E> Map<E, Set<AccountSkill>> appendSiteAndCorporateSkills(final Map<E, Set<AccountSkill>> entitySkillMap,
																	   final Collection<AccountSkill> siteAndCorporateRequiredSkills) {
		if (CollectionUtils.isEmpty(siteAndCorporateRequiredSkills)) {
			return entitySkillMap;
		}

		for (E entity : entitySkillMap.keySet()) {
			if (entitySkillMap.containsKey(entity)) {
				entitySkillMap.get(entity).addAll(siteAndCorporateRequiredSkills);
			}
		}

		return entitySkillMap;
	}

	public Set<Employee> getEmployeesAssignedToSite(final int siteId) {
		return getEmployeesAssignedToSites(Arrays.asList(siteId));
	}

	public Set<Employee> getEmployeesAssignedToSites(final Collection<Integer> siteIds) {
		List<Employee> employeesAssignedToSite = getSiteContractorEmployees(siteIds);

		final Map<Employee, Set<Project>> employeeProjects = projectEntityService.getProjectsForEmployeesBySiteIds(employeesAssignedToSite, siteIds);
		final Map<Employee, Set<Role>> employeeRoles = roleEntityService.getSiteRolesForEmployees(employeesAssignedToSite, siteIds);

		return new HashSet<Employee>() {{
			addAll(employeeProjects.keySet());
			addAll(employeeRoles.keySet());
		}};
	}

	private List<Employee> getSiteContractorEmployees(final Collection<Integer> siteIds) {
		List<Integer> contractorIds = accountService.getContractorIds(siteIds);
		return employeeEntityService.getEmployeesAssignedToSites(contractorIds, siteIds);
	}

	public Map<Employee, Set<Role>> getAllEmployeeRolesForSite(final int siteId) {
		List<Employee> employeesAssignedToSite = getSiteContractorEmployees(Arrays.asList(siteId));

		final Map<Employee, Set<Role>> employeeProjectRoles = roleEntityService.getProjectRolesForEmployees(employeesAssignedToSite, siteId);
		final Map<Employee, Set<Role>> employeeRoles = roleEntityService.getSiteRolesForEmployees(employeesAssignedToSite, siteId);

		return PicsCollectionUtil.mergeMapOfSets(employeeProjectRoles, employeeRoles);
	}

	public Map<Project, Set<Employee>> getEmployeesAssignedToProjects(final int siteId) {
		return getEmployeesAssignedToProjects(Arrays.asList(siteId));
	}

	public Map<Project, Set<Employee>> getEmployeesAssignedToProjects(final Collection<Integer> siteIds) {
		if (CollectionUtils.isEmpty(siteIds)) {
			return Collections.emptyMap();
		}

		Collection<Project> projects = projectEntityService.getProjectsBySiteIds(siteIds);
		return employeeEntityService.getEmployeesByProjects(projects);
	}

	public void assignEmployeeToProject(final Project project, final int roleId, final int employeeId,
										final EntityAuditInfo entityAuditInfo) {
		projectEntityService.assignEmployeeToProjectRole(project, roleId, employeeId, entityAuditInfo);

		assignEmployeeToSiteRole(project.getAccountId(), roleId, employeeId, entityAuditInfo);
	}

	public void unassignEmployeeFromProject(final Project project, final int roleId, final int employeeId) {
		projectEntityService.removeEmployeeFromProjectRole(project, roleId, employeeId);
	}

	public void assignEmployeeToSiteRole(final int siteId, final int roleId, final int employeeId,
										 final EntityAuditInfo entityAuditInfo) {
		roleEntityService.addEmployeeSiteAssignment(siteId, roleId, employeeId, entityAuditInfo);
	}

	public void unassignEmployeeFromSiteRole(final int siteId, final int roleId, final int employeeId) {
		roleEntityService.deleteEmployeeSiteAssignment(siteId, roleId, employeeId);

		projectEntityService.deleteEmployeeFromProjectRole(siteId, roleId, employeeId);
	}

	public void unassignEmployeeFromSite(final int siteId, final int employeeId) {
		roleEntityService.deleteAllEmployeeSiteAssignmentsForSite(siteId, employeeId);

		projectEntityService.unassignEmployeeFromAllProjectsOnSite(siteId, employeeId);
	}

	public Set<Integer> findAllEmployeeSiteAssignments(final Employee employee) {
		return employeeEntityService.getAllSiteIdsForEmployeeAssignments(employee);
	}
}
