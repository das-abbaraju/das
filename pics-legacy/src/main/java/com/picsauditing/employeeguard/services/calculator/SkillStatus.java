package com.picsauditing.employeeguard.services.calculator;

public enum SkillStatus {

    // TODO: Rename expiring to something more meaningful that the Skill is about to expire

    // Keep in order of serverity, from highest severity to lowest
    Expired("expired"),
    Expiring("expiring"),
    Pending("pending"),
    Complete("complete");

    private String displayValue;

    private SkillStatus(String displayValue) {
        this.displayValue = displayValue;
    }

    public boolean isExpired() {
        return this == Expired;
    }

    public boolean isExpiring() {
        return this == Expiring;
    }

    public boolean isPending() {
        return this == Pending;
    }

    public boolean isComplete() {
        return this == Complete;
    }

    public String getDisplayValue() {
        return displayValue;
    }
}
