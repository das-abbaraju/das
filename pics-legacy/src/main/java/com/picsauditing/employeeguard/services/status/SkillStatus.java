package com.picsauditing.employeeguard.services.status;

import com.picsauditing.jpa.entities.JSONable;
import org.json.simple.JSONObject;

// TODO: Rename so it is more generic since status is now applied to more than just skills
public enum SkillStatus implements JSONable {

    // TODO: Rename expiring to something more meaningful that the Skill is about to expire

    // Keep in order of severity, from highest severity to lowest
    Expired("expired"),    //-- Lowest Ordinal
    Expiring("expiring"),
    Pending("pending"),
    Completed("complete"); //-- Highest Ordinal

    private String displayValue;

    private SkillStatus(String displayValue) {
        this.displayValue = displayValue;
    }

    public boolean isExpired() {
        return this == Expired;
    }

    public boolean isExpiring() {
        return this == Expiring;
    }

    public boolean isPending() {
        return this == Pending;
    }

    public boolean isCompleted() {
        return this == Completed;
    }

    public String getDisplayValue() {
        return displayValue;
    }

	@Override
	public JSONObject toJSON(boolean full) {
		return new JSONObject();
	}

	@Override
	public void fromJSON(JSONObject o) {

	}
}
