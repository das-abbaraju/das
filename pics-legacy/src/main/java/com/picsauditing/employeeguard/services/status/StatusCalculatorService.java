package com.picsauditing.employeeguard.services.status;

import com.picsauditing.employeeguard.daos.AccountSkillProfileDAO;
import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.AccountSkillProfile;
import com.picsauditing.employeeguard.entities.Employee;
import com.picsauditing.employeeguard.entities.Profile;
import com.picsauditing.employeeguard.util.PicsCollectionUtil;
import com.picsauditing.util.generic.GenericPredicate;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.collections.MapUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

public class StatusCalculatorService {

	@Autowired
	private AccountSkillProfileDAO accountSkillProfileDAO;

	public Map<Employee, SkillStatus> getEmployeeStatusRollUpForSkills(final Map<Employee, Set<AccountSkill>> employeeSkills) {
		if (MapUtils.isEmpty(employeeSkills)) {
			return Collections.emptyMap();
		}

		Set<Employee> employees = employeeSkills.keySet();
		Map<Employee, Set<AccountSkillProfile>> accountSkillProfileMap =
				convertToMap(employees, accountSkillProfileDAO.findByEmployeesAndSkills(employees,
						PicsCollectionUtil.mergeCollectionOfCollections(employeeSkills.values())));

		Map<Employee, SkillStatus> employeeStatuses = new HashMap<>();
		for (Employee employee : employees) {
			SkillStatus rollUp = SkillStatus.Expired;

			Set<AccountSkillProfile> accountSkillProfiles = accountSkillProfileMap.get(employee);
			if (CollectionUtils.isNotEmpty(accountSkillProfiles)) {
				rollUp = SkillStatusCalculator.calculateStatusRollUp(accountSkillProfiles);
			}

			employeeStatuses.put(employee, rollUp);
		}

		return employeeStatuses;
	}

	@Deprecated
	public Map<Employee, SkillStatus> getEmployeeStatusRollUpForSkills(final Collection<Employee> employees,
																	   final Map<Employee, Set<AccountSkill>> employeeRequiredSkills) {
		if (CollectionUtils.isEmpty(employees) || MapUtils.isEmpty(employeeRequiredSkills)) {
			return Collections.emptyMap();
		}

		Map<Employee, SkillStatus> employeeStatuses = new HashMap<>();
		for (final Employee employee : employees) {
			SkillStatus rollUp = SkillStatus.Expired;
			if (CollectionUtils.isNotEmpty(employeeRequiredSkills.get(employee)) && employee.getProfile() != null) {
				List<AccountSkillProfile> employeeSkills = new ArrayList<>(employee.getProfile().getSkills());
				CollectionUtils.filter(employeeSkills, new GenericPredicate<AccountSkillProfile>() {

					@Override
					public boolean evaluateEntity(AccountSkillProfile accountSkillProfile) {
						return employeeRequiredSkills.get(employee).contains(accountSkillProfile.getSkill());
					}
				});

				if (CollectionUtils.isNotEmpty(employeeSkills)) {
					rollUp = SkillStatusCalculator.calculateStatusRollUp(employeeSkills);
				}
			}

			employeeStatuses.put(employee, rollUp);
		}

		return employeeStatuses;
	}

	/**
	 * Returns a Map where the Key is a unique set of every employee provided in Employees and the Value
	 * is an List contains a SkillStatus for every Skill in orderedSkills, in the same order.
	 *
	 * @param employees
	 * @param orderedSkills
	 * @return
	 */
	public Map<Employee, List<SkillStatus>> getEmployeeStatusRollUpForSkills(final Collection<Employee> employees,
																			 final List<AccountSkill> orderedSkills) {

		if (CollectionUtils.isEmpty(employees) || CollectionUtils.isEmpty(orderedSkills)) {
			return Collections.emptyMap();
		}

		List<AccountSkillProfile> accountSkillProfiles = accountSkillProfileDAO
				.findByEmployeesAndSkills(employees, orderedSkills);

		Map<Employee, Set<AccountSkillProfile>> employeeSetMap = convertToMap(employees, accountSkillProfiles);

		return buildSkillStatusMap(new HashSet<>(employees), employeeSetMap, orderedSkills);
	}

	private Map<Employee, Set<AccountSkillProfile>> convertToMap(final Collection<Employee> employees,
																 final List<AccountSkillProfile> accountSkillProfiles) {
		Map<Profile, Set<Employee>> employeeProfileMap = ServiceHelper.buildProfileToEmployeesMap(employees);

		Map<Profile, Set<AccountSkillProfile>> profileSkills = PicsCollectionUtil.convertToMapOfSets(accountSkillProfiles,

				new PicsCollectionUtil.MapConvertable<Profile, AccountSkillProfile>() {

					@Override
					public Profile getKey(AccountSkillProfile entity) {
						return entity.getProfile();
					}
				});

		return ServiceHelper.mapFromProfileToEmployee(profileSkills, employeeProfileMap);
	}

	private Map<Employee, List<SkillStatus>> buildSkillStatusMap(final Set<Employee> employees,
																 final Map<Employee, Set<AccountSkillProfile>> employeeMap,
																 final List<AccountSkill> orderedSkills) {

		final int numberOfSkills = orderedSkills.size();
		Map<Employee, List<SkillStatus>> employeeSkillStatusMap = new HashMap<>();
		for (Employee employee : employees) {
			if (employeeMap.containsKey(employee)) {
				employeeSkillStatusMap.put(employee, buildOrderedSkillStatusList(employeeMap.get(employee),
						orderedSkills));
			} else {
				employeeSkillStatusMap.put(employee, fillWithExpiredStatus(numberOfSkills));
			}
		}

		return employeeSkillStatusMap;
	}

	private List<SkillStatus> buildOrderedSkillStatusList(final Set<AccountSkillProfile> accountSkillProfiles,
														  final List<AccountSkill> orderedSkills) {
		List<SkillStatus> skillStatusList = fillWithExpiredStatus(orderedSkills.size());
		for (AccountSkillProfile accountSkillProfile : accountSkillProfiles) {
			int index = orderedSkills.indexOf(accountSkillProfile.getSkill());
			skillStatusList.set(index, SkillStatusCalculator.calculateStatusFromSkill(accountSkillProfile));
		}

		return skillStatusList;
	}

	private List<SkillStatus> fillWithExpiredStatus(final int size) {
		SkillStatus[] skillStatusArray = new SkillStatus[size];
		Arrays.fill(skillStatusArray, SkillStatus.Expired);
		return Arrays.asList(skillStatusArray);
	}

	public SkillStatus calculateOverallStatus(final Collection<SkillStatus> skillStatuses) {
		return SkillStatusCalculator.calculateOverallStatus(skillStatuses);
	}

	public <E> Map<E, SkillStatus> getOverallStatusPerEntity(final Map<E, List<SkillStatus>> entitySkillStatusMap) {
		return SkillStatusCalculator.getOverallStatusPerEntity(entitySkillStatusMap);
	}

	public <E> Map<E, List<SkillStatus>> getSkillStatusListPerEntity(final Employee employee, final Map<E, Set<AccountSkill>> skillMap) {
		if (employee == null || MapUtils.isEmpty(skillMap)) {
			return Collections.emptyMap();
		}

		Set<AccountSkill> skills = getSkillsFromMap(skillMap);
		if (CollectionUtils.isEmpty(skills)) {
			return Collections.emptyMap();
		}

		List<AccountSkillProfile> accountSkillEmployees = accountSkillProfileDAO
				.findByEmployeeAndSkills(employee, skills);
		Map<AccountSkill, AccountSkillProfile> accountSkillEmployeeMap =
				buildAccountSkillToAccountSkillEmployeeMap(accountSkillEmployees);

		return buildMapOfSkillStatus(skillMap, accountSkillEmployeeMap);
	}

	public <E> Map<E, List<SkillStatus>> getSkillStatusListPerEntity(final Employee employee,
																	 final Map<E, Set<AccountSkill>> skillMap,
																	 final SkillStatus defaultStatus) {
		if (employee == null || MapUtils.isEmpty(skillMap)) {
			return Collections.emptyMap();
		}

		Set<AccountSkill> skills = getSkillsFromMap(skillMap);
		if (CollectionUtils.isEmpty(skills)) {
			return Collections.emptyMap();
		}

		List<AccountSkillProfile> accountSkillEmployees = accountSkillProfileDAO
				.findByEmployeeAndSkills(employee, skills);

		Map<AccountSkill, AccountSkillProfile> accountSkillEmployeeMap =
				buildAccountSkillToAccountSkillEmployeeMap(accountSkillEmployees);

		return buildMapOfSkillStatus(skillMap, accountSkillEmployeeMap, defaultStatus);
	}

	public <E> Map<E, SkillStatus> getSkillStatusPerEntity(final Employee employee,
														   final Map<E, Set<AccountSkill>> skillMap,
														   final SkillStatus defaultStatus) {
		return getOverallStatusPerEntity(getSkillStatusListPerEntity(employee, skillMap, defaultStatus));
	}

	public <E> Map<E, SkillStatus> getSkillStatusPerEntity(final Employee employee, final Map<E, Set<AccountSkill>> skillMap) {
		return getOverallStatusPerEntity(getSkillStatusListPerEntity(employee, skillMap));
	}

	private <E> Set<AccountSkill> getSkillsFromMap(final Map<E, Set<AccountSkill>> skillMap) {
		if (MapUtils.isEmpty(skillMap)) {
			return Collections.emptySet();
		}

		Set<AccountSkill> skills = new HashSet<>();
		for (Set<AccountSkill> accountSkills : skillMap.values()) {
			skills.addAll(accountSkills);
		}

		return skills;
	}

	private Map<AccountSkill, AccountSkillProfile> buildAccountSkillToAccountSkillEmployeeMap(
			final List<AccountSkillProfile> accountSkillProfile) {

		return PicsCollectionUtil.convertToMap(accountSkillProfile,

				new PicsCollectionUtil.MapConvertable<AccountSkill, AccountSkillProfile>() {

					@Override
					public AccountSkill getKey(AccountSkillProfile accountSkillProfile) {
						return accountSkillProfile.getSkill();
					}
				});
	}

	private <E> Map<E, List<SkillStatus>> buildMapOfSkillStatus(final Map<E, Set<AccountSkill>> skillMap,
																final Map<AccountSkill, AccountSkillProfile> accountSkillEmployeeMap) {
		if (MapUtils.isEmpty(skillMap)) {
			return Collections.emptyMap();
		}

		Map<E, List<SkillStatus>> entityStatusMap = new HashMap<>();
		for (E entity : skillMap.keySet()) {
			List<SkillStatus> skillStatusList = new ArrayList<>();
			for (AccountSkill accountSkill : skillMap.get(entity)) {
				skillStatusList.add(SkillStatusCalculator.calculateStatusFromSkill(accountSkillEmployeeMap.get(accountSkill)));
			}

			entityStatusMap.put(entity, skillStatusList);
		}

		return entityStatusMap;
	}

	private <E> Map<E, List<SkillStatus>> buildMapOfSkillStatus(final Map<E, Set<AccountSkill>> skillMap,
																final Map<AccountSkill, AccountSkillProfile> accountSkillEmployeeMap,
																final SkillStatus defaultStatus) {
		if (MapUtils.isEmpty(skillMap)) {
			return Collections.emptyMap();
		}

		Map<E, List<SkillStatus>> entityStatusMap = new HashMap<>();
		for (E entity : skillMap.keySet()) {
			List<SkillStatus> skillStatusList = new ArrayList<>();
			if (CollectionUtils.isEmpty(skillMap.get(entity))) {
				skillStatusList.add(defaultStatus);
			} else {
				for (AccountSkill accountSkill : skillMap.get(entity)) {
					skillStatusList.add(SkillStatusCalculator.calculateStatusFromSkill(accountSkillEmployeeMap.get(accountSkill)));
				}
			}

			entityStatusMap.put(entity, skillStatusList);
		}

		return entityStatusMap;
	}

	public Map<AccountSkill, SkillStatus> getSkillStatuses(final Employee employee,
														   final Collection<AccountSkill> skills) {
		if (CollectionUtils.isEmpty(skills)) {
			return Collections.emptyMap();
		}

		List<AccountSkillProfile> accountSkillProfiles = accountSkillProfileDAO
				.findByEmployeeAndSkills(employee, skills);

		Map<AccountSkill, SkillStatus> skillStatuses = getAccountSkillStatusMap(accountSkillProfiles);
		return addExpiredStatusToSkillsEmployeeIsMissing(skills, skillStatuses);
	}

	private Map<AccountSkill, SkillStatus> addExpiredStatusToSkillsEmployeeIsMissing(final Collection<AccountSkill> skills,
																					 final Map<AccountSkill, SkillStatus> skillStatuses) {
		Map<AccountSkill, SkillStatus> allSkillStatuses = new HashMap<>();
		if (MapUtils.isNotEmpty(skillStatuses)) {
			allSkillStatuses.putAll(skillStatuses);
		}

		for (AccountSkill accountSkill : skills) {
			if (!allSkillStatuses.containsKey(accountSkill)) {
				allSkillStatuses.put(accountSkill, SkillStatus.Expired);
			}
		}

		return allSkillStatuses;
	}

	private Map<AccountSkill, SkillStatus> getAccountSkillStatusMap(List<AccountSkillProfile> accountSkillProfiles) {
		return PicsCollectionUtil.convertToMap(accountSkillProfiles,

				new PicsCollectionUtil.EntityKeyValueConvertable<AccountSkillProfile, AccountSkill, SkillStatus>() {
					@Override
					public AccountSkill getKey(AccountSkillProfile accountSkillProfile) {
						return accountSkillProfile.getSkill();
					}

					@Override
					public SkillStatus getValue(AccountSkillProfile accountSkillProfile) {
						return SkillStatusCalculator.calculateStatusFromSkill(accountSkillProfile);
					}
				});
	}

	public <E> Map<E, List<SkillStatus>> getAllSkillStatusesForEntity(final Map<E, Map<Employee, Set<AccountSkill>>> entityEmployeeSkillMap) {
		if (MapUtils.isEmpty(entityEmployeeSkillMap)) {
			return Collections.emptyMap();
		}

		///-- Prepare Unique list of employees and Skills
		Set<Employee> employees = new HashSet<>();
		Set<AccountSkill> skills = new HashSet<>();
		for (E entity : entityEmployeeSkillMap.keySet()) {
			for (Map.Entry<Employee, Set<AccountSkill>> entry : entityEmployeeSkillMap.get(entity).entrySet()) {
				employees.add(entry.getKey());
				skills.addAll(entry.getValue());
			}
		}

		//-- Get documentations for the list of employees with list of skills
		List<AccountSkillProfile> accountSkillProfile = Collections.emptyList();
		if (CollectionUtils.isNotEmpty(employees) && CollectionUtils.isNotEmpty(skills)) {
			accountSkillProfile = accountSkillProfileDAO.findByEmployeesAndSkills(employees, skills);
		}

    /* Prepare collection of Employees with skills that have documentation attached. This map of maps contains
	only the skills that the employee has provided documentation for !
    * */

		Map<Profile, Map<AccountSkill, AccountSkillProfile>> employeeSkillMap = PicsCollectionUtil.convertToMapOfMaps(

				accountSkillProfile,

				new PicsCollectionUtil.CollectionToMapConverter<Profile, AccountSkill, AccountSkillProfile>() {

					@Override
					public Profile getRow(AccountSkillProfile accountSkillProfile) {
						return accountSkillProfile.getProfile();
					}

					@Override
					public AccountSkill getColumn(AccountSkillProfile accountSkillProfile) {
						return accountSkillProfile.getSkill();
					}
				});

		//-- Roll up skill status of each Employee
		Map<E, List<SkillStatus>> skillStatusPerEntityEmployee = new HashMap<>();
		for (final E entity : entityEmployeeSkillMap.keySet()) {
			for (final Employee employee : entityEmployeeSkillMap.get(entity).keySet()) {

				SkillStatus skillStatus = SkillStatus.Expired;
				//-- If employee has any documentations at all.
				if (employeeSkillMap.containsKey(employee.getProfile())) {
					//-- Documentations provided for this employee
					Collection<AccountSkillProfile> aspForStatusCalculation = new HashSet<>(employeeSkillMap.get(employee.getProfile()).values());
					//-- Distill the collection specific to this employee's skills.
					CollectionUtils.filter(aspForStatusCalculation, new GenericPredicate<AccountSkillProfile>() {
						@Override
						public boolean evaluateEntity(AccountSkillProfile accountSkillProfile) {
							return entityEmployeeSkillMap.get(entity).get(employee).contains(accountSkillProfile.getSkill());
						}
					});

					if (!aspForStatusCalculation.isEmpty()) {
						skillStatus = SkillStatusCalculator.calculateStatusRollUp(aspForStatusCalculation);
					}
				}

				//-- Add Skill status information to the entity in question.
				if (!skillStatusPerEntityEmployee.containsKey(entity)) {
					skillStatusPerEntityEmployee.put(entity, new ArrayList<SkillStatus>());
				}
				skillStatusPerEntityEmployee.get(entity).add(skillStatus);

			}
		}

		return skillStatusPerEntityEmployee;
	}
}