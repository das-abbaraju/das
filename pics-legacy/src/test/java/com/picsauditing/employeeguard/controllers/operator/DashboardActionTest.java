package com.picsauditing.employeeguard.controllers.operator;

import com.picsauditing.PicsActionTest;
import com.picsauditing.actions.PicsActionSupport;
import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.Employee;
import com.picsauditing.employeeguard.entities.builders.EmployeeBuilder;
import com.picsauditing.employeeguard.services.AssignmentService;
import com.picsauditing.employeeguard.services.StatusCalculatorService;
import com.picsauditing.employeeguard.services.calculator.SkillStatus;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;

import java.util.HashMap;
import java.util.Set;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.Matchers.anyMap;
import static org.mockito.Matchers.anySetOf;
import static org.mockito.Mockito.when;

@UseReporter(DiffReporter.class)
public class DashboardActionTest extends PicsActionTest {
	public static final int CONTRACTOR_ID = 234;
	public static final int SITE_ID = 123;

	private DashboardAction dashboard;

	@Mock
	private AssignmentService assignmentService;
	@Mock
	private StatusCalculatorService statusCalculatorService;

	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);

		dashboard = new DashboardAction();
		super.setUp(dashboard);

		when(permissions.getAccountId()).thenReturn(SITE_ID);
		when(permissions.getAccountName()).thenReturn("Site Name");

		Whitebox.setInternalState(dashboard, "assignmentService", assignmentService);
		Whitebox.setInternalState(dashboard, "statusCalculatorService", statusCalculatorService);
	}

	@Test
	public void testIndex() throws Exception {
		when(assignmentService.getEmployeeSkillsForSite(SITE_ID)).thenReturn(new HashMap<Employee, Set<AccountSkill>>());
		when(statusCalculatorService.getEmployeeStatusRollUpForSkills(anySetOf(Employee.class), anyMap())).thenReturn(
				new HashMap<Employee, SkillStatus>() {{
					put(new EmployeeBuilder()
							.accountId(CONTRACTOR_ID)
							.firstName("Employee")
							.lastName("Test")
							.email("employee@test.com")
							.slug("EmployeeTest")
							.build(),
							SkillStatus.Complete);
					put(new EmployeeBuilder()
							.accountId(CONTRACTOR_ID)
							.firstName("Employee 2")
							.lastName("Test")
							.email("employee2@test.com")
							.slug("Employee2Test")
							.build(),
							SkillStatus.Expired);
				}});

		assertEquals(PicsActionSupport.JSON_STRING, dashboard.index());
		assertNotNull(dashboard.getJsonString());
		Approvals.verify(dashboard.getJsonString());
	}
}
