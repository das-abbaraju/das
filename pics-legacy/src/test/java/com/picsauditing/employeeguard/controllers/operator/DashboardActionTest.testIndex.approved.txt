{"sites":[{"id":123,"name":"Site Name","employees":2,"completed":1,"pending":0,"expiring":0,"expired":1,"projects":[]}]}